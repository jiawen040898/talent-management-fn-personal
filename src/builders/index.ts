export * from './base-query.builder';
export * from './employee-task-query.builder';
export * from './feedback-cycle-reviewee-query.builder';
export * from './feedback-cycle-reviewer-query.builder';
export * from './learning-goal-query.builder';
