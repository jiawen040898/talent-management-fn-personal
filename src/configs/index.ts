export * from './aws.config';
export * from './gcp.config';
export * from './recommendation.config';
