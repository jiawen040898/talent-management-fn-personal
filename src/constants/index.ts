export * from './action-source.enum';
export * from './big-queries.enum';
export * from './cache-store.enum';
export * from './company-module.enum';
export * from './company-status.enum';
export * from './due-type.enum';
export * from './employee-score.enum';
export * from './employee-task.enum';
export * from './event-type.enum';
export * from './feedback-recommendation.enum';
export * from './fit-score-recipe.enum';
export * from './person-score.enum';
export * from './psychology-api-errors.enum';
export * from './questionnaire-framework-type.enum';
export * from './sqs-event-type.enum';
export * from './unleash-errors.enum';
