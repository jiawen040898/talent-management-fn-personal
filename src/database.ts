import { secretService } from '@pulsifi/fn';
import { DataSource } from 'typeorm';

import { ENTITIES } from './models';

type DbSecret = {
    username: string;
    engine: string;
    port: number;
    host: string;
    password: string;
    schema: string;
    dbname: string;
    dbInstanceIdentifier: string;
};

let dataSource: DataSource;

export const getDataSource = async (): Promise<DataSource> => {
    if (dataSource?.isInitialized) {
        return dataSource;
    }

    const dbSecret = await secretService.getSecret<DbSecret>(
        process.env.SM_NAME as string,
    );

    dataSource = await new DataSource({
        type: 'postgres',
        host: dbSecret.host,
        port: dbSecret.port,
        username: dbSecret.username,
        password: dbSecret.password,
        database: dbSecret.dbname,
        schema: dbSecret.schema,
        entities: [...ENTITIES],
        maxQueryExecutionTime: 1000,
        synchronize: false,
        migrationsRun: false,
        migrations: [],
        extra: {
            connectionTimeoutMillis: 3000,
            idleTimeoutMillis: 1000,
            max: 1,
        },
        logging: ['error'],
    }).initialize();

    return dataSource;
};
