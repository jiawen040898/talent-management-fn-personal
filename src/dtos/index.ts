export * from './company.dto';
export * from './cron-schedule-message.dto';
export * from './employee.dto';
export * from './employee-account-created-event.dto';
export * from './employee-assessment-score-event.dto';
export * from './employee-fit-score-event.dto';
export * from './feedback-cycle.dto';
export * from './fit-score-recipe.dto';
export * from './manager-subordinate-aggregated-report-data.dto';
