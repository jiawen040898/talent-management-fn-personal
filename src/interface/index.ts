export * from '../dtos/fit-score-recipe.dto';
export * from './cognitive-score.interface';
export * from './employee-score.interface';
export * from './event-model.interface';
export * from './feedback-cycle.interface';
export * from './feedback-recommendation.interface';
export * from './fit-score.interface';
export * from './personality-score.interface';
export * from './program.interface';
export * from './redis.interface';
