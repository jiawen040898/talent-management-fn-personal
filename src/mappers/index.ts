export * from './employee-score.mapper';
export * from './feedback.mapper';
export * from './feedback-recommendation.mapper';
