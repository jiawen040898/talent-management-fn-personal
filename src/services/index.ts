export * from './cognitive-score.service';
export * from './company.service';
export * from './domain-event-router.service';
export * from './employee-daily-reminder.service';
export * from './employee-management.service';
export * from './employee-progress-report.service';
export * from './employee-score.service';
export * from './feedback-cycle-action.service';
export * from './feedback-cycle-dashboard-provision.service';
export * from './feedback-cycle-reminder.service';
export * from './feedback-cycle-reviewee-recommendation.service';
export * from './fit-score.service';
export * from './personality-score.service';
export * from './psychology.service';
export * from './recommendation.service';
