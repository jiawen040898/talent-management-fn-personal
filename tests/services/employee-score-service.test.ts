import { SNSClient } from '@aws-sdk/client-sns';
import { dateTimeUtil } from '@pulsifi/fn';
import { DataSource } from 'typeorm';

import * as database from '../../src/database';
import { EmployeeScoreService } from '../../src/services';
import { testData } from '../services/fixtures/test-data';
import { getTestDataSource, getTestDataSourceAndAddData } from '../setup';
import { TestData } from '../setup/test-data.setup';
import { cultureFitRecipeTestData } from './fixtures/culture-fit-recipe.test-data';
import { employeeScoreTestData } from './fixtures/employee-score-service.test-data';
import { roleFitRecipeTestData } from './fixtures/role-fit-recipe.test-data';

jest.mock('../../src/services/psychology.service', () => {
    const mBInstance = {
        getFitScoreRecipeById: jest.fn(
            () => testData.mockGetFitScoreRecipe.data.data,
        ),
    };
    const mB = jest.fn(() => mBInstance);
    return { PsychologyService: mB };
});

const mockDateTimeNow = jest.spyOn(dateTimeUtil, 'now');
mockDateTimeNow.mockReturnValue(new Date('2023-01-31T12:34:56.789Z'));
const mockGetFitScoreRecipe = jest.fn();

const spySNSClient = jest
    .spyOn(SNSClient.prototype, 'send')
    .mockImplementation(jest.fn());

describe('[employeeScoreService]', () => {
    let dataSource: DataSource;
    const spyGetDataSource = jest.spyOn(database, 'getDataSource');

    beforeAll(async () => {
        dataSource = await getTestDataSource();
        EmployeeScoreService.getFitScoreRecipe = mockGetFitScoreRecipe;

        dataSource = await getTestDataSourceAndAddData(
            employeeScoreTestData.entitiesToBeAdded,
        );
        spyGetDataSource.mockImplementation(() => {
            return Promise.resolve(dataSource);
        });
    });

    it('should pass processPersonalityScore', async () => {
        // Act
        const result = await EmployeeScoreService.processPersonality(
            employeeScoreTestData.personalityAssessmentResults,
            employeeScoreTestData.participant,
            roleFitRecipeTestData.validRoleFitScoreRecipe,
            roleFitRecipeTestData.validRoleFitScoreRecipe.id,
        );

        // Assert
        expect(result).toMatchSnapshot();
    });

    it('should pass processCognitive', async () => {
        // Act
        const result = await EmployeeScoreService.processCognitive(
            employeeScoreTestData.cognitiveAssessmentResults,
            employeeScoreTestData.participant,
            roleFitRecipeTestData.validRoleFitScoreRecipe.id,
        );

        // Assert
        expect(result).toMatchSnapshot();
    });

    describe('processEmployeeAssessmentScoreCalculated', () => {
        beforeEach(() => {
            spySNSClient.mockClear();
        });

        it('should pass sendTMAssessmentScoreCalculated event', async () => {
            // Act
            await EmployeeScoreService.processEmployeeAssessmentScoreCalculated(
                employeeScoreTestData.participant,
                cultureFitRecipeTestData.validCultureFitScoreRecipe,
                TestData.companyId,
            );

            // Assert
            expect(spySNSClient).toHaveBeenCalledTimes(1);
        });
    });

    describe('processEmployeeAssessmentScore', () => {
        const roleFitRecipeId =
            roleFitRecipeTestData.validRoleFitScoreRecipe.id;
        const cultureFitRecipeId =
            cultureFitRecipeTestData.validCultureFitScoreRecipe.id;

        beforeEach(() => {
            mockGetFitScoreRecipe.mockClear();
            mockGetFitScoreRecipe.mockImplementation((id) =>
                id === roleFitRecipeId
                    ? roleFitRecipeTestData.validRoleFitScoreRecipe
                    : cultureFitRecipeTestData.validCultureFitScoreRecipe,
            );
            spySNSClient.mockClear();
        });

        it.each([
            [
                'role fit id and cognitive assessment result',
                roleFitRecipeId,
                undefined,
                employeeScoreTestData.cognitiveAssessmentResults,
                1,
                0,
            ],
            [
                'role fit id and personality assessment result',
                roleFitRecipeId,
                undefined,
                employeeScoreTestData.personalityAssessmentResults,
                1,
                0,
            ],
            [
                'culture fit id and cognitive assessment result',
                undefined,
                cultureFitRecipeId,
                employeeScoreTestData.cognitiveAssessmentResults,
                1,
                1,
            ],
            [
                'role fit id, culture fit id and cognitive assessment result',
                roleFitRecipeId,
                cultureFitRecipeId,
                employeeScoreTestData.cognitiveAssessmentResults,
                2,
                1,
            ],
            [
                'role fit id, culture fit id and personality assessment result',
                roleFitRecipeId,
                cultureFitRecipeId,
                employeeScoreTestData.personalityAssessmentResults,
                2,
                1,
            ],
            [
                'role fit id, culture fit id and all assessment result',
                roleFitRecipeId,
                cultureFitRecipeId,
                [
                    ...employeeScoreTestData.personalityAssessmentResults,
                    ...employeeScoreTestData.cognitiveAssessmentResults,
                ],
                2,
                1,
            ],
        ])(
            'should pass if there are %s in employee assessment submitted event',
            async (
                _,
                roleFitRecipeId,
                cultureFitRecipeId,
                assessments,
                expectGetFitScoreRecipeToBeCalledTimes,
                expectSnsToBeCalledTimes,
            ) => {
                // Act
                await EmployeeScoreService.processEmployeeAssessmentScore({
                    company_id: TestData.companyId,
                    program_id: employeeScoreTestData.participant.program_id,
                    participant_id: employeeScoreTestData.participant.id,
                    role_fit_recipe_id: roleFitRecipeId,
                    culture_fit_recipe_id: cultureFitRecipeId,
                    assessments: assessments,
                });

                // Assert
                expect(mockGetFitScoreRecipe).toHaveBeenCalledTimes(
                    expectGetFitScoreRecipeToBeCalledTimes,
                );
                expect(spySNSClient).toHaveBeenCalledTimes(
                    expectSnsToBeCalledTimes,
                );
            },
        );
    });
});
