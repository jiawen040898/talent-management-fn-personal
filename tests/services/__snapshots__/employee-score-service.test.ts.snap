// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[employeeScoreService] should pass processCognitive 1`] = `
[
  {
    "created_by": 1,
    "employee_id": "00000000-0000-0000-0000-000000000001",
    "id": 0,
    "participant_id": "00000000-0000-0000-0000-000000000001",
    "score": 13.043478260869565,
    "score_dimension": 0,
    "score_outcome": {
      "cognitive_result": {
        "domain_alias": "numeric",
        "domain_percentile": null,
        "domain_score": 0.13043478260869565,
        "ingredient_weightage": null,
      },
    },
    "score_recipe_id": "00000000-0000-0000-0000-000000000001",
    "score_type": "reasoning_numeric",
    "updated_by": 1,
  },
  {
    "created_by": 1,
    "employee_id": "00000000-0000-0000-0000-000000000001",
    "id": 2,
    "participant_id": "00000000-0000-0000-0000-000000000001",
    "score": 26.666666666666668,
    "score_dimension": 0,
    "score_outcome": {
      "cognitive_result": {
        "domain_alias": "verbal",
        "domain_percentile": null,
        "domain_score": 0.26666666666666666,
        "ingredient_weightage": null,
      },
    },
    "score_recipe_id": "00000000-0000-0000-0000-000000000001",
    "score_type": "reasoning_verbal",
    "updated_by": 1,
  },
  {
    "created_by": 1,
    "employee_id": "00000000-0000-0000-0000-000000000001",
    "id": 1,
    "participant_id": "00000000-0000-0000-0000-000000000001",
    "score": 27.77777777777778,
    "score_dimension": 0,
    "score_outcome": {
      "cognitive_result": {
        "domain_alias": "logical",
        "domain_percentile": null,
        "domain_score": 0.2777777777777778,
        "ingredient_weightage": null,
      },
    },
    "score_recipe_id": "00000000-0000-0000-0000-000000000001",
    "score_type": "reasoning_logical",
    "updated_by": 1,
  },
  {
    "created_by": 1,
    "employee_id": "00000000-0000-0000-0000-000000000001",
    "id": undefined,
    "participant_id": "00000000-0000-0000-0000-000000000001",
    "score": 22.49597423510467,
    "score_dimension": 0,
    "score_outcome": {
      "cognitive_result": {
        "domain_alias": "reasoning_average",
        "domain_percentile": null,
        "domain_score": 0.22495974235104668,
        "ingredient_weightage": null,
      },
    },
    "score_recipe_id": "00000000-0000-0000-0000-000000000001",
    "score_type": "reasoning_average",
    "updated_by": 1,
  },
]
`;

exports[`[employeeScoreService] should pass processPersonalityScore 1`] = `
[
  {
    "created_by": 1,
    "employee_id": "00000000-0000-0000-0000-000000000001",
    "id": 3,
    "participant_id": "00000000-0000-0000-0000-000000000001",
    "score": 7.451600000000001,
    "score_dimension": 0,
    "score_outcome": {
      "personality_result": [
        {
          "domain_alias": "achievement_effort",
          "domain_score": 0.8229166666666666,
          "domain_weightage": 0.06271626297577854,
          "traits": [],
        },
        {
          "domain_alias": "persistence",
          "domain_score": 0.8125,
          "domain_weightage": 0.06459054209919263,
          "traits": [],
        },
        {
          "domain_alias": "initiative",
          "domain_score": 0.7678571428571429,
          "domain_weightage": 0.06891580161476356,
          "traits": [],
        },
        {
          "domain_alias": "leadership",
          "domain_score": 0.7458333333333333,
          "domain_weightage": 0.0658881199538639,
          "traits": [],
        },
        {
          "domain_alias": "cooperation",
          "domain_score": 0.75,
          "domain_weightage": 0.06271626297577854,
          "traits": [],
        },
        {
          "domain_alias": "concern_for_others",
          "domain_score": 0.625,
          "domain_weightage": 0.05392156862745098,
          "traits": [],
        },
        {
          "domain_alias": "social_orientation",
          "domain_score": 0.8125,
          "domain_weightage": 0.05521914648212226,
          "traits": [],
        },
        {
          "domain_alias": "self_control",
          "domain_score": 0.8375,
          "domain_weightage": 0.061995386389850055,
          "traits": [],
        },
        {
          "domain_alias": "stress_tolerance",
          "domain_score": 0.7291666666666666,
          "domain_weightage": 0.0651672433679354,
          "traits": [],
        },
        {
          "domain_alias": "adaptability_flexibility",
          "domain_score": 0.828125,
          "domain_weightage": 0.0658881199538639,
          "traits": [],
        },
        {
          "domain_alias": "dependability",
          "domain_score": 0.625,
          "domain_weightage": 0.0658881199538639,
          "traits": [],
        },
        {
          "domain_alias": "attention_to_detail",
          "domain_score": 0.625,
          "domain_weightage": 0.0589677047289504,
          "traits": [],
        },
        {
          "domain_alias": "integrity",
          "domain_score": 0.8125,
          "domain_weightage": 0.06689734717416378,
          "traits": [],
        },
        {
          "domain_alias": "independence",
          "domain_score": 0.6666666666666666,
          "domain_weightage": 0.061995386389850055,
          "traits": [],
        },
        {
          "domain_alias": "innovation",
          "domain_score": 0.8125,
          "domain_weightage": 0.059544405997693194,
          "traits": [],
        },
        {
          "domain_alias": "analytical_thinking",
          "domain_score": 0.625,
          "domain_weightage": 0.05968858131487889,
          "traits": [],
        },
      ],
    },
    "score_recipe_id": "00000000-0000-0000-0000-000000000001",
    "score_type": "work_style",
    "updated_by": 1,
  },
  {
    "created_by": 1,
    "employee_id": "00000000-0000-0000-0000-000000000001",
    "id": 4,
    "participant_id": "00000000-0000-0000-0000-000000000001",
    "score": 6.770722000000001,
    "score_dimension": 0,
    "score_outcome": {
      "personality_result": [
        {
          "domain_alias": "achievement",
          "domain_score": 0.8373015873015873,
          "domain_weightage": 0.17475409836065575,
          "traits": [],
        },
        {
          "domain_alias": "independence",
          "domain_score": 0.6956349206349207,
          "domain_weightage": 0.1859016393442623,
          "traits": [],
        },
        {
          "domain_alias": "recognition",
          "domain_score": 0.7083333333333333,
          "domain_weightage": 0.15311475409836064,
          "traits": [],
        },
        {
          "domain_alias": "relationships",
          "domain_score": 0.545,
          "domain_weightage": 0.13114754098360656,
          "traits": [],
        },
        {
          "domain_alias": "support",
          "domain_score": 0.5861111111111111,
          "domain_weightage": 0.17475409836065575,
          "traits": [],
        },
        {
          "domain_alias": "working_conditions",
          "domain_score": 0.6603174603174603,
          "domain_weightage": 0.18032786885245902,
          "traits": [],
        },
      ],
    },
    "score_recipe_id": "00000000-0000-0000-0000-000000000001",
    "score_type": "work_value",
    "updated_by": 1,
  },
  {
    "created_by": 1,
    "employee_id": "00000000-0000-0000-0000-000000000001",
    "id": 5,
    "participant_id": "00000000-0000-0000-0000-000000000001",
    "score": 8.5714,
    "score_dimension": 0,
    "score_outcome": {
      "job_codes": [
        "E",
        "C",
        "S",
      ],
      "person_codes": [
        [
          "S",
          "C",
          "E",
        ],
      ],
      "personality_result": [
        {
          "domain_alias": "realistic",
          "domain_score": 0.3125,
        },
        {
          "domain_alias": "investigative",
          "domain_score": 0.3125,
        },
        {
          "domain_alias": "artistic",
          "domain_score": 0.28125,
        },
        {
          "domain_alias": "social",
          "domain_score": 0.8125,
        },
        {
          "domain_alias": "enterprising",
          "domain_score": 0.71875,
        },
        {
          "domain_alias": "conventional",
          "domain_score": 0.78125,
        },
      ],
    },
    "score_recipe_id": "00000000-0000-0000-0000-000000000001",
    "score_type": "work_interest",
    "updated_by": 1,
  },
]
`;
